File: src/ru/meldren/functionalprogramming/lab1/Task20.java
package ru.meldren.functionalprogramming.lab1;

import java.math.BigInteger;

public class Task20 {

    public int digitSum(BigInteger n) {
        int sum = 0;
        String str = n.toString();
        for (int i = 0; i < str.length(); i++) {
            sum += str.charAt(i) - '0';
        }
        return sum;
    }

    private BigInteger findFactorial(int n) {
        BigInteger factorial = BigInteger.ONE;
        for (int i = 1; i <= n; i++) {
            factorial = factorial.multiply(BigInteger.valueOf(i));
        }
        return factorial;
    }
}

File: src/ru/meldren/functionalprogramming/lab1/task20.clj
(ns ru.meldren.functionalprogramming.lab1.task20)

;; Факториал с использованием рекурсии
(defn factorial-rec [n]
  (if (= n 1)
    1
    (* n (factorial-rec (dec n)))))

;; Факториал с использованием хвостовой рекурсии
(defn factorial-tail-rec-acc [n acc]
  (if (= n 1)
    acc
    (recur (dec n) (* n acc))))

(defn factorial-tail-rec [n]
  (factorial-tail-rec-acc n 1))

;; Факториал с использованием свертки
(defn factorial-reduce [n]
  (reduce *' (range 1 (inc n))))

;; Факториал с использованием ленивых вычислений
(defn factorial-lazy [n]
  (reduce *' (take n (iterate inc 1))))

;; Функция для нахождения суммы цифр числа
(defn digit-sum [n]
  (->> (str n)
       (map #(Character/digit ^char % 10))
       (reduce +)))
